sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/StringTypeCollated.scala

```diff
@@ -52,3 +52,20 @@ case object StringTypeAnyCollation extends StringTypeCollated {
   override private[sql] def simpleString: String = "string_any_collation"
   override private[sql] def acceptsType(other: DataType): Boolean = other.isInstanceOf[StringType]
 }
+
+/**
+ * Use ArrayType(StringTypeCollated) for expressions supporting different type of
+ * collations in ArrayTypes
+ */
+case class AbstractArrayType(elementType: StringTypeCollated) extends AbstractDataType {
```

Comment:
can we make it general to take `AbstractDataType`? I don't think we should limit it to string type.

Reply:
Done, I also realised, should I move the StringTypeCollated to the right package? It is now in catalyst.expressions instead of types as all other types.